apps/emqx_cluster_link/src/emqx_cluster_link_bookkeeper.erl

```diff
@@ -0,0 +1,84 @@
+%%--------------------------------------------------------------------
+%% Copyright (c) 2024 EMQ Technologies Co., Ltd. All Rights Reserved.
+%%--------------------------------------------------------------------
+-module(emqx_cluster_link_bookkeeper).
+
+%% API
+-export([
+    start_link/0
+]).
+
+%% `gen_server' API
+-export([
+    init/1,
+    handle_continue/2,
+    handle_call/3,
+    handle_cast/2,
+    handle_info/2
+]).
+
+%%------------------------------------------------------------------------------
+%% Type declarations
+%%------------------------------------------------------------------------------
+
+%% call/cast/info events
+-record(tally_routes, {}).
+
+%%------------------------------------------------------------------------------
+%% API
+%%------------------------------------------------------------------------------
+
+-spec start_link() -> gen_server:start_ret().
+start_link() ->
+    gen_server:start_link({local, ?MODULE}, ?MODULE, _InitOpts = #{}, _Opts = []).
+
+%%------------------------------------------------------------------------------
+%% `gen_server' API
+%%------------------------------------------------------------------------------
+
+init(_Opts) ->
+    State = #{},
+    {ok, State, {continue, #tally_routes{}}}.
+
+handle_continue(#tally_routes{}, State) ->
+    handle_tally_routes(),
+    {noreply, State}.
+
+handle_call(_Call, _From, State) ->
+    {reply, {error, bad_call}, State}.
+
+handle_cast(_Cast, State) ->
+    {noreply, State}.
+
+handle_info(#tally_routes{}, State) ->
+    handle_tally_routes(),
+    {noreply, State};
+handle_info(_Info, State) ->
+    {noreply, State}.
+
+%%------------------------------------------------------------------------------
+%% Internal fns
+%%------------------------------------------------------------------------------
+
+cluster_names() ->
+    Links = emqx_cluster_link_config:links(),
+    lists:map(fun(#{name := Name}) -> Name end, Links).
+
+ensure_timer(Event, Timeout) ->
+    _ = erlang:send_after(Timeout, self(), Event),
+    ok.
+
+handle_tally_routes() ->
+    ClusterNames = cluster_names(),
+    tally_routes(ClusterNames),
+    ensure_timer(#tally_routes{}, emqx_cluster_link_config:tally_routes_interval()),
+    ok.
+
+tally_routes([ClusterName | ClusterNames]) ->
+    Tab = emqx_cluster_link_extrouter:extroute_tab(),
+    Pat = emqx_cluster_link_extrouter:cluster_routes_ms(ClusterName),
+    NumRoutes = ets:select_count(Tab, Pat),
+    emqx_cluster_link_metrics:routes_set(ClusterName, NumRoutes),
```

Comment:
Just basically to kick off a discussion: having so "full-scan-intensive" job just to have lagging but accurate metrics feels a bit wasteful to me. I wonder if we can have a hybrid approach, where (as before) a counter tracks per-node route additions and deletions (accepting the risk of slight drift under high concurrency) + this bookkeeper that tracks cluster-wide gauges, but it runs much more rarely and resets the per-node counters on each run. Then `NRoutes(ClusterName) = Gauge(ClusterName) + Sum(Counter(Node, ClusterName) || N <- Nodes)`. Bonus point is almost no lagging, downside is lack of 100% accuracy.

I'm not sure though how it fits into the current metrics models, please share any thoughts. Maybe I'm wrong about this perceived wastefulness?

Reply:
Ah wait, it won't work as I've imagined, counter resets will happen at different times on different nodes, which will deform the aggregation results.

Reply:
Maybe that could work.  Each node would keep track of its own counters and have those fast but potentially inaccurate counters that would be reset by the bookkeeper.  The difference though is that we should have `NRoutes(ClusterName) = Gauge(ClusterName) + Counter(ClusterName)` done on each node, and then we can take the maximum value when aggregating.  Again, there's room for lagging metrics or a bit of inaccuracy, but sounds reasonable.

The problem I can think of now is what you mentioned before about having potentially more than 1 actor inserting the same route...  Not sure how we could easily deduplicate that.  :thinking:

Reply:
> The problem I can think of now is what you mentioned before about having potentially more than 1 actor inserting the same route... Not sure how we could easily deduplicate that. 🤔

Yeah I meant that we should essentially mirror the (former) delete logic, i.e.:
```
    case MCounter band Marker of
        0 when OpName =:= add ->
            case mria:dirty_update_counter(?EXTROUTE_TAB, Entry, Marker) of
                Marker -> %% (likely) we're first, bump metric.
                _More -> %% not the first actor to push that route.
            end;
        Marker when OpName =:= add ->
            %% Already added.
            MCounter;
        Marker when OpName =:= delete ->
            case mria:dirty_update_counter(?EXTROUTE_TAB, Entry, -Marker) of
                0 ->
                    %% (likely) the last actor to delete that route, decrease the metric
                    Record = #extroute{entry = Entry, mcounter = 0},
                    ok = mria:dirty_delete_object(?EXTROUTE_TAB, Record),
                    0;
                C ->
                    %% there are still actors holding onto this route
                    C
            end;
        0 when OpName =:= delete ->
            %% Already deleted.
            MCounter
    end.
```

Reply:
> The difference though is that we should have `NRoutes(ClusterName) = Gauge(ClusterName) + Counter(ClusterName)` done on each node, and then we can take the maximum value when aggregating.

Won't it be a bit off because counters are still per-node, so even the maximum will not reflect changes made by _all_ actors (which can be connected to different nodes)? I.e. won't something like this will be closer to the truth?
```
[G1 | Gauges] = [Gauge(ClusterName @ N) || N <- Nodes],
[M1 | Metrics] = [Counter(ClusterName @ N) || N <- Nodes],
G1 + M1 + sum([G - G1 + M || {G, M} <- zip(Gauges, Metrics)])
```

It's obvious this starts to look complex and hard to reason about, so might be a subject for a followup. 🫠